### ISSUE TITLE
Train Killed after several epochs

### ISSUE THREAD

**Issue by EdwinGerman** [itdxer/neupy#203]
Hello Dears:
I am training a model using k-folds cross-validation( k= 30 and 150000 epochs). The Database used has 5000 rows and 37 features. After two days of training the process stop and only the message: "Killed" at the command prompt. If you have any idea about what is happening it would help a lot.

I am using a GEFORCE GTX 1080 Ti and my computer has 16GB of RAM to train the model.
 
**The message:**

>   3.5857e+06 │    0.044099 │    0.030027 │       23 ms │
> │  3.5858e+06 │    0.042621 │    0.022126 │       23 ms │
> │  3.5859e+06 │    0.047204 │   0.0077744 │       23 ms │
> │   3.586e+06 │    0.056358 │    0.011478 │       23 ms │
> │  3.5861e+06 │    0.045348 │    0.010573 │       23 ms │
> │  3.5862e+06 │    0.050718 │    0.024256 │       23 ms │
> │  3.5863e+06 │    0.047565 │   0.0065777 │       23 ms │
> │  3.5864e+06 │    0.044024 │   0.0083317 │       23 ms │
> │  3.5865e+06 │    0.056907 │    0.010707 │       23 ms │
> │  3.5866e+06 │    0.050813 │   0.0060741 │       23 ms │
> │  3.5867e+06 │    0.061978 │   0.0078684 │       23 ms │
> │  3.5868e+06 │    0.048344 │    0.010392 │       23 ms │
> │  3.5869e+06 │    0.050844 │   0.0077159 │       23 ms │
> │   3.587e+06 │    0.042654 │     0.01516 │       26 ms │
> │  3.5871e+06 │    0.050458 │    0.063693 │       23 ms │
> │  3.5872e+06 │    0.049051 │   0.0060742 │       23 ms │
> │  3.5873e+06 │    0.048567 │   0.0075477 │       23 ms │
> Killed

**The code:**

```
from neupy.exceptions import StopTraining
from neupy import algorithms, layers, storage
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from neupy import plots
import pandas as pd
import  math
import os
from sklearn.model_selection import KFold


MAX_ROUNDS   = 150
tested_round =0
WINNER_EPOCH =0


def average(x):
    assert len(x) > 0
    return float(sum(x)) / len(x)

def pearson_def(x, y):
    assert len(x) == len(y)
    n = len(x)
    assert n > 0
    avg_x = average(x)
    avg_y = average(y)
    diffprod = 0
    xdiff2 = 0
    ydiff2 = 0
    for idx in range(n):
        xdiff = x[idx] - avg_x
        ydiff = y[idx] - avg_y
        diffprod += xdiff * ydiff
        xdiff2 += xdiff * xdiff
        ydiff2 += ydiff * ydiff

    return diffprod / math.sqrt(xdiff2 * ydiff2)

def on_epoch_end1(adamnet):
    global tested_round
    epoch = adamnet.last_epoch
    errors = adamnet.validation_errors

    if errors.previous() and errors.last() > errors.previous():
        # Load parameters and stop training
        storage.load(adamnet, 'training-epoch-{}.pickle'.format(epoch - 1))
        raise StopTraining("Training has been interrupted")
    else:
        # Save parameters after successful epoch
        storage.save(adamnet, 'training-epoch-{}.pickle'.format(epoch))
    tested_round = tested_round+1

def on_epoch_end(adamnet):
    global  tested_round
    global  WINNER_EPOCH
    epoch = adamnet.last_epoch
    tested_round1 = tested_round
    #errors = adamnet.validation_errors
    validation_error = adamnet.validation_errors
    #train_error      = adamnet.train_errors
    #tested_round =0
    if validation_error.previous() is None:
        previous_value = 0
    else:
        previous_value = validation_error.previous()
    if  validation_error.last() > previous_value:
       if tested_round1==0:
           WINNER_EPOCH = epoch
       tested_round1    = tested_round1 + 1
    else:
        if abs( validation_error.last() - previous_value )>0.00001:
            tested_round1 = tested_round1 + 1
        else:
            # Save parameters after successful epoch
            storage.save(adamnet, 'training-epoch-{}.pickle'.format(epoch))
            tested_round1 = 0
            #winner_epoch  = epoch
    if tested_round==MAX_ROUNDS:
        print("Winner :" + str(WINNER_EPOCH))
        storage.load(adamnet, 'training-epoch-{}.pickle'.format(WINNER_EPOCH-1))
        raise StopTraining("Training has been interrupted")
    tested_round = tested_round1
    print(tested_round1)


    # if errors.previous() and errors.last() > errors.previous():
    #     # Load parameters and stop training
    #     storage.load(adamnet, 'training-epoch-{}.pickle'.format(epoch - 1))
    #     raise StopTraining("Training has been interrupted")
    # else:
    #     # Save parameters after successful epoch
    #     storage.save(adamnet, 'training-epoch-{}.pickle'.format(epoch))

percentaje_test = 0.10



def main(database_path, bool_k_fold, num_epochs, n_splits=10):
    full_dataset = pd.read_csv(database_path)
    # full_dataset = full_dataset.ix[:,[0, 1, 3, 5, 6, 8, 10]].reset_index(drop=True)
    # full_dataset = full_dataset.ix[:,[0, 1, 3, 5, 8, 10]].reset_index(drop=True)
    full_dt_rows = full_dataset.shape[0]
    n_rows_train = full_dt_rows - int(full_dt_rows * percentaje_test)
    n_cols_train = full_dataset.shape[1]

    x_train_dataset = full_dataset.ix[0:n_rows_train, 0:n_cols_train - 1].reset_index(drop=True)
    y_train_dataset = full_dataset.ix[0:n_rows_train, n_cols_train - 1].reset_index(drop=True)

    x_scaler = preprocessing.RobustScaler().fit(x_train_dataset)
    y_scaler = preprocessing.StandardScaler().fit(y_train_dataset.values.reshape(-1, 1))

    x_train_scaled = x_scaler.transform(x_train_dataset)
    y_train_scaled = y_scaler.transform(y_train_dataset.values.reshape(-1, 1))

    x_test_dataset = full_dataset.ix[n_rows_train:full_dataset.shape[0], 0:n_cols_train - 1].reset_index(drop=True)
    y_test_dataset = full_dataset.ix[n_rows_train:full_dataset.shape[0], n_cols_train - 1].reset_index(drop=True)

    x_test_scaled  = x_scaler.transform(x_test_dataset)
    # y_test_scaled        = y_scaler.transform(y_test_dataset.reshape(-1, 1))

    # train_dataset = full_dataset.ix[0:n_rows_train,:].reset_index(drop=True) #train + validation
    # test_dataset  = full_dataset.ix[n_rows_train:full_dt_rows,:].reset_index(drop=True) #independent test


    # scaler_train = preprocessing.StandardScaler().fit(train_dataset)

    # train_dataset_scaled = scaler_train.transform(train_dataset)

    # n_cols = train_dataset_scaled.shape[1]
    # data   = x_train_scaled
    # target = train_dataset_scaled.ix[:,n_cols-1]
    # target = y_train_scaled
    # data   = data.as_matrix()
    # n_cols_test = test_dataset.shape[1]
    # data_test   = x_test_scaled
    # target_test = y_test_scaled
    # data_test   = data_test.as_matrix()
    # target_test = target_test.as_matrix()
    # dataset = datasets.load_boston()
    # data, target = dataset.data, dataset.target
    # data_scaler   = preprocessing.MinMaxScaler()
    # target_scaler = preprocessing.MinMaxScaler()
    # data   = data_scaler.fit_transform(data)
    # target = target_scaler.fit_transform(target.reshape(-1, 1))
    # environment.reproducible()
    Adamnet = algorithms.Adamax(
        [layers.Input(x_train_scaled.shape[1]),

	 layers.BatchNorm(),
         layers.Tanh(10),
         layers.Dropout(0.01),
	 
	 #layers.BatchNorm(),
         #layers.Tanh(30),
         #layers.Dropout(0.5),

	 #layers.BatchNorm(),
         #layers.Tanh(30),
         #layers.Dropout(0.5),

	 layers.BatchNorm(),
         layers.Tanh(10),
         layers.Dropout(0.01),

	 layers.BatchNorm(),
         layers.Linear(1), ],

         shuffle_data = True,
         batch_size   = 100,
         verbose      = True,
         show_epoch   = 100,
         epsilon      = 1e-8,
    )
    if bool_k_fold:
        kf = KFold(n_splits=n_splits)
        kf.get_n_splits(x_train_scaled)
        for train_index, test_index in kf.split(x_train_scaled):
            x_train, x_val = x_train_scaled[train_index], x_train_scaled[test_index]
            y_train, y_val = y_train_scaled[train_index], y_train_scaled[test_index]
            Adamnet.train(x_train, y_train, x_val, y_val, epochs=num_epochs)
    else:
        x_train, x_val, y_train, y_val = train_test_split(
            x_train_scaled, y_train_scaled, train_size=0.90, shuffle=True,
        )
        Adamnet.train(x_train, y_train, x_val, y_val, epochs=num_epochs)

    plots.error_plot(Adamnet)

    with Adamnet.connection.disable_training_state():
        y_predicted = Adamnet.predict(x_test_scaled)

    #y_predicted        = Adamnet.predict(x_test_scaled)
    y_predicted_denorm = y_scaler.inverse_transform(y_predicted)

    R1 = pearson_def(y_predicted_denorm, y_test_dataset)

    print(R1)

if __name__ == "__main__":
    #experiment_number = 30
    full_database_name= "fisa_casp3_full_treated.csv" #"fisa_casp3_full.csv" #fisa_casp3_full_non_duplicate2.csv"
    database_path     = os.path.join(os.sep,"home","edwin","workspace1",full_database_name)
    bool_k_fold       = True
    n_splits_folds    = 30

    #Neural Network Parameters
    num_epochs       = 150000#150000

    main(database_path, bool_k_fold, num_epochs,n_splits=n_splits_folds)
```
Thank you for your help.

**Comment by itdxer** [comment]
Hi,
It's hard to tell, in this case it would be even hard to debug. i have to think about this issue, I'm not sure what could be the problem. It looks like issue with memory, but I'm not sure.

One thing that I noticed from the logs is that table shows epoch number greater than 150,000 (the last one was 3,587,300).

```python
        for train_index, test_index in kf.split(x_train_scaled):
            x_train, x_val = x_train_scaled[train_index], x_train_scaled[test_index]
            y_train, y_val = y_train_scaled[train_index], y_train_scaled[test_index]
            Adamnet.train(x_train, y_train, x_val, y_val, epochs=num_epochs)
```

In the code you train the same network after each iteration. I think that before training you have to re-initialize new network and reset progress from the previous iteration.

I would also suggest you, just for safety reasons, saving networks from time to time, especially if you run for a long time. Check: http://neupy.com/docs/storage.html

**Comment by EdwinGerman** [comment]
Thank you very much for your answers.  Yes, I am training the same network in each split of my database. In k-fold cross-validation, I think we don't need to reset the network in each fold of training. Am I wrong? or is necessary to do it in NeuPy (when k-folds cross-validation is used).

The idea of saving my model from time to time is great I will put it in the next version of this code.

**Comment by itdxer** [comment]
On the first kfold split you get different training and test data and using training data you train the network. On the next kfold split part of training data from the previous kfold will be in test and at this point validation won't show you real prediction power of the network since it has already seen this data during the training.

**Comment by itdxer** [comment]
Also I think that for 5000 data points 150,000 iterations can be too much. I think even 1,000 can be a lot for this case. You might also consider to reduce number of epochs in order to speed up your training.

**Comment by EdwinGerman** [comment]
Thank you for your advice. Well at the beginning of the simulation I separate the database into two sets: 1)Train  2)Test. The first, I divide in: a)Train  b)Validation. These two sets are used to find the best model with the k-fold cross-validation.The validation set is only used to check the generalization of the model, this set is not used to assess the performance of the final model.

When the training process has finished the Test set is used to compare the predicted values and the real values(Pearson correlation). The Test set has points never seen by the model because they are separated at the beginning of the process. 

I confess I am a bit confused about how to apply k-folds cross-validation. I see some codes like:
[](http://thelillysblog.com/2017/08/18/machine-learning-k-fold-validation/)
[](http://scikit-learn.org/stable/auto_examples/exercises/plot_cv_diabetes.html)
In these examples, I did not see that they need to reset the model in each split of the database.


Yes, you are right the number of epochs is too much for this database. My idea is to do a stress test in the reduced one and then run this program over the whole database. My real database has 534, 000 data points.

**Comment by itdxer** [comment]
> I confess I am a bit confused about how to apply k-folds cross-validation. I see some codes like:

In the first article they re-train new classifier in each iteration  (http://thelillysblog.com/2017/08/18/machine-learning-k-fold-validation/)

```python
for train, test in kf.split(data):
    ...
    classifier = nltk.NaiveBayesClassifier.train(train_data)
```

In scikit-learn example  (http://scikit-learn.org/stable/auto_examples/exercises/plot_cv_diabetes.html) it's a bit more tricky. Basically, every call of the `fit` function, for most of the algorithms in scikit-learn, will force to train model from scratch. Which means that in the example when they trigger `fit` they re-train from scratch completely ignoring knowledge from the previous training. In neupy calling `train` multiple times will just continue training procedure.

> My real database has 534, 000 data points.

I see, but in case of bigger datasets you will need even less iterations for training since you show more examples per each epoch. Now you use 5,000 example, which means that after 100 iterations you will show 500,000 samples to the network and if you have dataset with 500,000 data points then you will show the same number of samples per one epoch.

**Comment by itdxer** [comment]
NeuPy has been moved to tensorflow and I have to close this issue since it was related to the theano library which was completely removed starting from 0.7.0 version
